federation/federation.go

```diff
@@ -0,0 +1,195 @@
+package federation
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+var (
+	headSnapshotKey = datastore.NewKey("head-snapshot")
+	logger          = logging.Logger("federation")
+)
+
+type (
+	Federation struct {
+		*options
+		members []peer.ID
+		vc      vectorClock
+
+		mu                   sync.Mutex
+		snapshotTicker       *time.Ticker
+		reconciliationTicker *time.Ticker
+		headNodeCache        ipld.Node
+		ctx                  context.Context
+		cancel               context.CancelFunc
```

Comment:
Prefer letting the caller cancel rather than storing on the struct?

Reply:
this is to cancel the event loop started internally by Start. 

We can use the context passed into start. but that context might be only used for starting the event loop and could be surprising to the users if it is tied to the entire lifecyle of the federation.